src/kOS.Safe/Compilation/KS/Compiler.cs

```diff
@@ -413,76 +413,85 @@ private void PreProcessChildNodes(ParseNode node)
+            AddOpcode(new OpcodePush(triggerObject.OldValueIdentifier));
+            AddOpcode(new OpcodeSwap());
+            AddOpcode(new OpcodeStoreGlobal());
+            // Use the other dup'ed copy of the new value to actually do the equals
+            // comparison with the old value that's still under it on the stack:
             AddOpcode(new OpcodeCompareEqual());
-            AddOpcode(new OpcodeLogicNot());
-            Opcode branchOpcode = AddOpcode(new OpcodeBranchIfFalse());
+            OpcodeBranchIfFalse branchToBody = new OpcodeBranchIfFalse();
+            branchToBody.Distance = 3;
+            AddOpcode(branchToBody);
+            AddOpcode(new OpcodePush(true));       // wasn't triggered yet, so preserve.
+            AddOpcode(new OpcodeReturn((short)0)); // premature return because it wasn't triggered
 
             // make flag that remembers whether to remove trigger:
             // defaults to true = removal should happen.
-            string triggerRemoveVarName = "$remove-" + triggerIdentifier;
-            PushTriggerRemoveName(triggerRemoveVarName);
-            AddOpcode(new OpcodePush(triggerRemoveVarName));
-            AddOpcode(new OpcodePush(true));
+            string triggerKeepName = "$keep-" + triggerIdentifier;
+            PushTriggerKeepName(triggerKeepName);
+            AddOpcode(new OpcodePush(triggerKeepName));
+            AddOpcode(new OpcodePush(false));
             AddOpcode(new OpcodeStoreGlobal());
 
             VisitNode(node.Nodes[2]);
 
-            // Reset the "old value" so the boolean has to change *again* to re-trigger
-            AddOpcode(new OpcodePush(triggerObject.VariableNameOldValue));
-            AddOpcode(new OpcodePush(triggerObject.VariableName));
-            AddOpcode(new OpcodeStore());
-
-            // Skip removing the trigger if PRESERVE happened:
-            PopTriggerRemoveName(); // Throw away return value.
-            AddOpcode(new OpcodePush(triggerRemoveVarName));
-            Opcode skipRemoval = AddOpcode(new OpcodeBranchIfFalse());
-
-            AddOpcode(new OpcodePushRelocateLater(null), triggerObject.GetFunctionLabel());
-            AddOpcode(new OpcodeRemoveTrigger());
-            Opcode eofOpcode = AddOpcode(new OpcodeEOF());
-            branchOpcode.DestinationLabel = eofOpcode.Label;
-            skipRemoval.DestinationLabel = eofOpcode.Label;
+            // PRESERVE will determine whether or not the trigger returns true (true means
+            // re-enable the trigger upon exit.)
+            PopTriggerKeepName();
+            AddOpcode(new OpcodePush(triggerKeepName));
+            AddOpcode(new OpcodeReturn((short)0));
+            
+            nextBraceIsFunction = false;
         }
 
         private void PreProcessWhenStatement(ParseNode node)
         {
             NodeStartHousekeeping(node);
+            nextBraceIsFunction = true; // triggers aren't really functions but act like it a lot.
+            
```

Comment:
This change was to support having the ``return`` keyword work properly within the trigger's body.